src/Mobs/Villager.cpp

```diff
@@ -195,6 +374,107 @@ void cVillager::HandleFarmerPlaceCrops()
 
 
 
+void cVillager::UpdateTradeTier(unsigned DesiredTier)
+{
+	struct CareerHash
+	{
+		size_t operator()(const VillagerCareer & Career) const
+		{
+			return static_cast<size_t>(Career);
+		}
+	};
+	using Offer = VillagerTradeOffer;
+
+	static const std::unordered_map<
+		VillagerCareer,
+		std::vector<
+			// Element index: trading tier (0-based)
+			std::vector<Offer>  // Trade offer
+		>,
+		CareerHash  // TODO for C++14: unneeded
+	> TradeMatrix =
+	{
+		{
+			VillagerCareer::Farmer,
+			{
+				{  // Tier 1 (equivalent)
+					Offer(cItem(E_ITEM_WHEAT, 18), cItem(E_ITEM_EMERALD)),
+					Offer(cItem(E_ITEM_POTATO, 15), cItem(E_ITEM_EMERALD)),
+					Offer(cItem(E_ITEM_CARROT, 15), cItem(E_ITEM_EMERALD))
+				},
+				{  // Tier 2
+					Offer(cItem(E_BLOCK_PUMPKIN, 8), cItem(E_ITEM_EMERALD)),
+					Offer(cItem(E_ITEM_EMERALD), cItem(E_ITEM_PUMPKIN_PIE, 2))
+				},
+				{  // Tier 3
+					Offer(cItem(E_BLOCK_MELON, 7), cItem(E_ITEM_EMERALD)),
+					Offer(cItem(E_ITEM_EMERALD), cItem(E_ITEM_RED_APPLE, 5))
+				},
+				{  // Tier 4
+					Offer(cItem(E_ITEM_EMERALD), cItem(E_ITEM_COOKIE, 6)),
+					Offer(cItem(E_ITEM_EMERALD), cItem(E_ITEM_CAKE))
+				}
+			}
+		},
+		{
+			VillagerCareer::Fisherman, {}
+		},
+		{
+			VillagerCareer::Shepherd, {}
+		},
+		{
+			VillagerCareer::Fletcher, {}
+		},
+		{
+			VillagerCareer::Librarian, {}
+		},
+		{
+			VillagerCareer::Cartographer, {}
+		},
+		{
+			VillagerCareer::Cleric, {}
+		},
+		{
+			VillagerCareer::Armorer, {}
+		},
+		{
+			VillagerCareer::WeaponSmith, {}
+		},
+		{
+			VillagerCareer::ToolSmith, {}
+		},
+		{
+			VillagerCareer::Butcher, {}
+		},
+		{
+			VillagerCareer::Leatherworker, {}
+		},
+		{
+			VillagerCareer::Nitwit, {}
+		}
+	};
+
+	auto Tiers = TradeMatrix.find(m_Career)->second;
+	if (Tiers.empty())
+	{
+		// (This case is not handled 'naturally' since the tier variables are unsigned)
+		// (Otherwise the loop below would just exit immediately, rather elegantly. But alas ¯\_(ツ)_/¯)
+		return;
+	}
+
+	m_TradeTier = std::min(static_cast<unsigned>(Tiers.size() - 1), DesiredTier);
+	m_TradeOffers.clear();
+
+	for (unsigned Tier = 0; Tier <= m_TradeTier; Tier++)
+	{
+		m_TradeOffers.insert(m_TradeOffers.end(), Tiers[Tier].cbegin(), Tiers[Tier].cend());
+	}
```

Comment:
You could use the inclusive-exclusive [0, `m_TradeTier`) range like the STL.  Then the empty case would be handled naturally.  Another implication being you could do `UpdateTradeTier(0)` to disable trading.